tips/TIP-0018/tip-0018.md

```diff
@@ -0,0 +1,3755 @@
+
+Outputs are records in the UTXO ledger that track ownership of funds. Thus, each output must be able to specify which
+funds it holds. With the addition of the Native Tokenization Framework, outputs may also carry user defined native
+tokens, that is, tokens that are not IOTA coins but were minted by foundries and are tracked in the very same ledger.
+Therefore, **every output must be able to hold not only IOTA coins, but also native tokens**.
+
+Dust protection applies to all outputs, therefore it is not possible for outputs to hold only native tokens, the dust
+requirements must be covered via IOTA coins.
+
+User defined tokens are called <i>Native Tokens</i> on protocol level. The maximum supply of a particular native token
+is defined by the representation chosen on protocol level for defining their amounts in outputs. Since native tokens
+are also a vehicle to wrap layer 2 tokens into layer 1 tokens, the chosen representation must take into account the
+maximum possible supply of layer 2 tokens. Solidity, the most popular layer 2 smart contract language defines the
+maximum supply of an ERC-20 token as `MaxUint256`, therefore it should be possible to represent such huge amount of
+assets on layer 1.
+
+Outputs must have the following fields to define the balance of native tokens they hold:
+
+<table>
+    <tr>
+        <td><b>Name</b></td>
+        <td><b>Type</b></td>
+        <td><b>Description</b></td>
+    </tr>
+    <tr>
+        <td>Native Tokens Count</td>
+        <td>uint16</td>
+        <td>The number of native tokens present in the output.</td>
+    </tr>
+    <tr>
+        <td valign="top">Native Tokens <code>optAnyOf</code></td>
+        <td colspan="2">
+            <details>
+                <summary>Native Token</summary>
+                <table>
+                    <tr>
+                        <td><b>Name</b></td>
+                        <td><b>Type</b></td>
+                        <td><b>Description</b></td>
+                    </tr>
+                    <tr>
+                        <td>Token ID</td>
+                        <td>ByteArray[38]</td>
+                        <td>
+                            Identifier of the native token. Derivation defined <a href=https://github.com/lzpap/protocol-rfcs/blob/master/tips/TIP-0018/tip-0018.md#foundry-output>here</a>.
+                        </td>
+                    </tr>
+                    <tr>
+                        <td>Amount</td>
+                        <td>uint256</td>
+                        <td>
+                            Amount of tokens.
+                        </td>
+                    </tr>
+                </table>
+            </details>
+        </td>
+    </tr>
+</table>
+
+#### Additional syntactic output validation rules:
+
+- `Native Tokens` must be lexicographically sorted based on `Token ID`.
+- Each <i>Native Token</i> must be unique in the set of `Native Tokens` based on its `Token ID`. No duplicates are
+  allowed.
```

Comment:
Shouldn't we syntactically check that the amount of any native token in the set is not 0 ?

Reply:
Yes we should, thanks!